[Spring 체스 - 1단계] 스컬(김윤기) 미션 제출합니다.

skullkim opened the pull request:
루피! 안녕히세요! 리뷰 잘부탁해요ㅎㅎ.
스프링이 처음이라 많이 해맨 미션이였어요.. 아직도 뭐가 뭔지 잘 모르겠어서 많이 공부해야 할거 같아요..

Ohzzi committed:
feat: 기존 자바 체스 이식

Ohzzi committed:
feat: 메인 페이지 컨트롤러에 매핑

Ohzzi committed:
feat: 게임 방 번호를 입력하면 게임으로 입장되도록 URL 매핑 구현

Ohzzi committed:
feat: 게임 정보를 불러오는 api 기능과 URL 매핑 완료

Ohzzi committed:
chore: 불필요한 라이브러리 의존성 제거

Ohzzi committed:
test: WebController 페이지 렌더링 테스트 추가

Ohzzi committed:
refactor: GameDao가 JdbcTemplate을 사용하도록 변경

Ohzzi committed:
refactor: @Transactional 어노테이션을 추가하여 deleteAll 메서드를 대체

Ohzzi committed:
fix: 데이터베이스 설정 수정

Ohzzi committed:
refactor: DAO가 DataSource만 주입받도록 변경

Ohzzi committed:
refactor: PieceDao가 JdbcTemplate을 사용하도록 변경

Ohzzi committed:
refactor: Service를 Spring의 Service로 변경

Ohzzi committed:
refactor: 게임을 시작하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 게임을 재시작하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 체스 기물을 움직이는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 점수를 출력하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 게임을 종료하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 예외 발생 시 예외 메시지를 클라이언트로 보내주는 기능을 Spring Controller로 이식

Ohzzi committed:
fix: mysql 설정 오류 수정

Ohzzi committed:
chore: 불필요한 파일 및 의존성 제거

Ohzzi committed:
fix: 여러 테스트를 통합으로 진행 시 테이블 중복 생성이 되어 테스트가 되지 않는 오류 수정

Ohzzi committed:
refactor: 체스 보드를 만드는 메서드 분리

Ohzzi committed:
refactor: 게임 조회와 생성 로직 분리

Ohzzi committed:
refactor: 게임 시작 및 재시작 로직을 한 메서드와 URL로 통합

Ohzzi committed:
refactor: 체스 기물 이동 URL 및 메서드 변경

Ohzzi committed:
refactor: 게임 상태 조회 기능 URL 변경

Ohzzi committed:
refactor: POST 요청 성공 시 201 코드 반환하도록 변경

Ohzzi committed:
refactor: 게임 조회 시 게임이 없으면 404 NOT FOUND 반환하도록 변경

Ohzzi committed:
refactor: api URL 간소화

Ohzzi committed:
refactor: Service에서 안쓰는 메서드 제거

workromancer commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,72 @@
+    public ChessGameResponse endGame(@PathVariable long gameId) {
+        return chessService.end(gameId);
+    }
+    // TODO: 게임 삭제 기능으로 변경 필요
```
주석은 삭제해도 괜찮지 않을까요 ㅎㅎ

skullkim replied:
수정하겠습니다.

workromancer commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @PostMapping("/{gameId}")
+    public ResponseEntity<ChessGameResponse> createGame(@PathVariable long gameId) throws URISyntaxException {
```
checked exception, unchecked exception 의 처리 대해서 고민해보는 것도 좋을 것 같아요

skullkim replied:
checked exception을 사용하고 다른 메서드에서 해당 checked exception을 catch 해야 한다면 두 메서드 사이에서 호출되는 모든 메서드들에 throws를 명시해야 하죠. 따라서 throws를 명시한 메서들은 checked exception이 발생할 수 있는 메서드의 내부 구현을 알고 있는 것이고 그에 따라 캡슐화가 깨지는 것으로 알고 있습니다. 
그래서 저도 평소에서는 checked exception 대신 unchecked exception을 주로 사용합니다. 만약 불가피하게 checked exception이 발생하는 경우 곧바로 catch해 unchecked exceptino으로 변환해 throw를 하고요. 하지만 이 부분의 경우 컨트롤러에서 throws를 하고 있고 ExceptionHandler가 곧장 이 에러를 처리하고 있기 떄문에 굳이 이 부분에 대해서도 unchecked exception으로 변환해 주어야 할지 의문입니다.
나중에 다른 메서드가 해당 메서드를 호출할 경우를 대비한다면 unchecked exception을 사용하는 것이 맞는거 같습니다. 하지만 컨트롤러에서 매핑한 메서드여서 이 매서드를 다른 매서드가 호출할 경우는 없는거 같습니다.
결론적으로 스프링의 컨트롤러에서는 checked exception을 사용해도 무방하다고 생각합니다.

workromancer commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,72 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
👍

skullkim replied:
😁

workromancer commented on src/main/java/chess/controller/WebController.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.web.servlet.view.RedirectView;
+
+@Controller
+public class WebController {
```
Controller 와 RestController 를 명확한 기준으로 잘 나눠주신 것 같아요

Controller 는 static file 응답만, RestController 는 dynamic response 만 해주고 계신 것으로 파악했어요 👍

skullkim replied:
😁

workromancer commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,72 @@
+
+    private final ChessService chessService;
+
+    public ChessController(ChessService chessService) {
```
생성자 주입은 어노테이션이 없어도 되는 이유는 다음 코멘트를 참고해주세요 https://github.com/woowacourse/jwp-chess/pull/330#discussion_r856673339

skullkim replied:
감사합니다~

workromancer commented on src/main/java/chess/dao/GameDaoImpl.java:
```diff
@@ -0,0 +1,64 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class GameDaoImpl implements GameDao {
```
Impl 보다 명확한 네이밍은 없을까요 ?

skullkim replied:
수정하겠습니다.

workromancer commented on src/test/java/chess/dao/PieceDaoTest.java:
```diff
@@ -0,0 +1,101 @@
+import chess.domain.piece.Color;
+import chess.domain.piece.Pawn;
+import chess.domain.piece.Piece;
+import com.mysql.cj.exceptions.AssertionFailedException;
```
아래 이미지 설정의 `Run tests using` 을 Gradle 로 변경해서 실행해보시고 실패하는 이유를 학습해보시는 것도 좋을 것 같아요.
아마 Intellij 로 설정해서 모든 테스트가 통과하는 것으로 보이는데요.

프로젝트를 빌드할 때는 gradle 기반으로 테스트를 돌리기 때문에 시간이 되신다면 학습해보시는 것도 좋을 것 같아요.

<img width="982" alt="image" src="https://user-images.githubusercontent.com/52121827/165003612-3749ab12-7a6e-4bed-8848-7ca43f45dc17.png">

workromancer reviewed (approved):
스컬 안녕하세요~ 리뷰어 루피입니다.

스프링 처음이신 것이 맞나요 ?! 전반적으로 잘 구현해주셨어요.
참고해주시면 좋을만한 내용들 코멘트로 남겨두었어요.

이번 미션은 스프링 적용하는 것을 체험해보는 것이라서, 2단계로 넘어가도 좋을 것 같아요.
머지하고 다음 step2 에서 뵐께요 !

workromancer commented:
@skullkim 확인되시면 아래 체크박스에 체크해주세요 

- [x] 실행 관련 확인중

skullkim committed:
fix(game.js): 새로운 아이디로 개임 시작 안되는 문제 해결

잘못된 응답 처리로 인해 새로운 게임을 처리하지 못하는 문제 해결.
새로운 개임을 생성할 경우 응답엔 json이 없지만 생성 여부를 확인하기 전에
json을 파싱하려해 문제가 발생

BREAKING CHANGE: 응답 처리 로직 변경
as-is:
window.onload = async function () {

  ...

  const res = await fetch(`/games/${gameId}`);
  if (!res.ok) {
    await create();
    return;
  }
  const data = await res.json();
  load(data);
}

window.onload = async function () {

  ...

  const res = await fetch(`/games/${gameId}`);
  const data = await res.json();
  if (!res.ok) {
    await create();
    return;
  }
  load(data);
}

skullkim committed:
chore(ChessController.java): 불필요한 주석 삭제
